File: src/shared/types/offer-brief.ts
import { z } from 'zod';

// ─── Enumerations ────────────────────────────────────────────────────────────

export type TriggerType = 'marketer_initiated' | 'purchase_triggered';
export type OfferStatus = 'draft' | 'approved' | 'active' | 'expired';

export const TRIGGER_TYPES: TriggerType[] = ['marketer_initiated', 'purchase_triggered'];
export const OFFER_STATUSES: OfferStatus[] = ['draft', 'approved', 'active', 'expired'];

// ─── Nested Type Interfaces ──────────────────────────────────────────────────

export interface Segment {
  name: string;
  definition: string;
  estimated_size: number;
  criteria: string[];
}

export interface Construct {
  type: string;
  value: number;
  description: string;
}

export interface Channel {
  channel_type: 'push' | 'email' | 'sms' | 'in_app';
  priority: number;
  message_template?: string;
}

export interface KPIs {
  expected_redemption_rate: number;
  expected_uplift_pct: number;
  target_segment_size?: number;
}

export interface RiskFlags {
  over_discounting: boolean;
  cannibalization: boolean;
  frequency_abuse: boolean;
  offer_stacking: boolean;
  severity: 'low' | 'medium' | 'critical';
  warnings: string[];
}

// ─── Core OfferBrief Interface ────────────────────────────────────────────────

export interface OfferBrief {
  offer_id: string;
  objective: string;
  segment: Segment;
  construct: Construct;
  channels: Channel[];
  kpis: KPIs;
  risk_flags: RiskFlags;
  status: OfferStatus;
  trigger_type: TriggerType;
  created_at: string; // ISO 8601
  valid_until?: string; // ISO 8601 — only for purchase_triggered offers
}

// ─── Inventory Suggestion (AI mode) ──────────────────────────────────────────

export interface InventorySuggestion {
  product_id: string;
  product_name: string;
  category: string;
  store: string;
  units_in_stock: number;
  urgency: 'high' | 'medium' | 'low';
  suggested_objective: string;
  stale?: boolean;
}

// ─── Zod Schemas ─────────────────────────────────────────────────────────────

export const SegmentSchema = z.object({
  name: z.string().min(1, 'Segment name is required'),
  definition: z.string().min(1, 'Segment definition is required'),
  estimated_size: z.number().int().nonnegative(),
  criteria: z.array(z.string()).min(1, 'At least one segment criterion is required'),
});

export const ConstructSchema = z.object({
  type: z.string().min(1, 'Construct type is required'),
  value: z.number().nonnegative(),
  description: z.string().min(1, 'Construct description is required'),
});

export const ChannelSchema = z.object({
  channel_type: z.enum(['push', 'email', 'sms', 'in_app']),
  priority: z.number().int().min(1),
  message_template: z.string().optional(),
});

export const KPIsSchema = z.object({
  expected_redemption_rate: z.number().min(0).max(1),
  expected_uplift_pct: z.number().min(0),
  target_segment_size: z.number().int().nonnegative().optional(),
});

export const RiskFlagsSchema = z.object({
  over_discounting: z.boolean(),
  cannibalization: z.boolean(),
  frequency_abuse: z.boolean(),
  offer_stacking: z.boolean(),
  severity: z.enum(['low', 'medium', 'critical']),
  warnings: z.array(z.string()),
});

export const OfferBriefSchema = z.object({
  offer_id: z.string().uuid('offer_id must be a valid UUID'),
  objective: z
    .string()
    .min(10, 'Objective must be at least 10 characters')
    .max(500, 'Objective must not exceed 500 characters'),
  segment: SegmentSchema,
  construct: ConstructSchema,
  channels: z.array(ChannelSchema).min(1, 'At least one channel is required'),
  kpis: KPIsSchema,
  risk_flags: RiskFlagsSchema,
  status: z.enum(['draft', 'approved', 'active', 'expired']),
  trigger_type: z.enum(['marketer_initiated', 'purchase_triggered']),
  created_at: z.string().datetime({ message: 'created_at must be ISO 8601' }),
  valid_until: z.string().datetime({ message: 'valid_until must be ISO 8601' }).optional(),
});

// ─── Input Schema for Offer Generation ───────────────────────────────────────

export const GenerateOfferInputSchema = z.object({
  objective: z
    .string()
    .min(10, 'Objective must be at least 10 characters')
    .max(500, 'Objective must not exceed 500 characters'),
  segment_hints: z.array(z.string()).optional(),
});

export type GenerateOfferInput = z.infer<typeof GenerateOfferInputSchema>;
export type OfferBriefValidated = z.infer<typeof OfferBriefSchema>;

File: src/backend/models/offer_brief.py
"""Pydantic v2 models for OfferBrief — mirrors src/shared/types/offer-brief.ts exactly."""

from __future__ import annotations

from datetime import datetime
from enum import Enum
from typing import Optional
from uuid import UUID

from pydantic import BaseModel, Field, model_validator


class TriggerType(str, Enum):
    marketer_initiated = "marketer_initiated"
    purchase_triggered = "purchase_triggered"


class OfferStatus(str, Enum):
    draft = "draft"
    approved = "approved"
    active = "active"
    expired = "expired"


class ChannelType(str, Enum):
    push = "push"
    email = "email"
    sms = "sms"
    in_app = "in_app"


class RiskSeverity(str, Enum):
    low = "low"
    medium = "medium"
    critical = "critical"


# ─── Nested Models ────────────────────────────────────────────────────────────


class Segment(BaseModel):
    name: str = Field(..., min_length=1)
    definition: str = Field(..., min_length=1)
    estimated_size: int = Field(..., ge=0)
    criteria: list[str] = Field(..., min_length=1)


class Construct(BaseModel):
    type: str = Field(..., min_length=1)
    value: float = Field(..., ge=0)
    description: str = Field(..., min_length=1)


class Channel(BaseModel):
    channel_type: ChannelType
    priority: int = Field(..., ge=1)
    message_template: Optional[str] = None


class KPIs(BaseModel):
    expected_redemption_rate: float = Field(..., ge=0, le=1)
    expected_uplift_pct: float = Field(..., ge=0)
    target_segment_size: Optional[int] = Field(default=None, ge=0)


class RiskFlags(BaseModel):
    over_discounting: bool
    cannibalization: bool
    frequency_abuse: bool
    offer_stacking: bool
    severity: RiskSeverity
    warnings: list[str] = Field(default_factory=list)


# ─── Core OfferBrief Model ────────────────────────────────────────────────────


class OfferBrief(BaseModel):
    offer_id: str = Field(..., description="UUID v4 string")
    objective: str = Field(..., min_length=10, max_length=500)
    segment: Segment
    construct: Construct
    channels: list[Channel] = Field(..., min_length=1)
    kpis: KPIs
    risk_flags: RiskFlags
    status: OfferStatus = OfferStatus.draft
    trigger_type: TriggerType = TriggerType.marketer_initiated
    created_at: datetime = Field(default_factory=datetime.utcnow)
    valid_until: Optional[datetime] = None

    model_config = {"from_attributes": True}

    @model_validator(mode="after")
    def validate_valid_until_for_purchase_triggered(self) -> "OfferBrief":
        """Purchase-triggered offers must have valid_until set."""
        if self.trigger_type == TriggerType.purchase_triggered and self.valid_until is None:
            raise ValueError(
                "valid_until is required for purchase_triggered offers"
            )
        return self


# ─── Request / Response Models ────────────────────────────────────────────────


class GenerateOfferRequest(BaseModel):
    objective: str = Field(..., min_length=10, max_length=500)
    segment_hints: Optional[list[str]] = None

    model_config = {
        "json_schema_extra": {
            "example": {
                "objective": "Reactivate lapsed high-value members this winter",
                "segment_hints": ["high_value", "lapsed_90_days"],
            }
        }
    }


class ApproveOfferResponse(BaseModel):
    offer_id: str
    status: OfferStatus
    hub_saved: bool
    message: str


class FraudCheckResult(BaseModel):
    severity: RiskSeverity
    flags: RiskFlags
    warnings: list[str]
    blocked: bool


class InventorySuggestion(BaseModel):
    product_id: str
    product_name: str
    category: str
    store: str
    units_in_stock: int = Field(..., ge=0)
    urgency: str = Field(..., pattern="^(high|medium|low)$")
    suggested_objective: str
    stale: bool = False


class DealSuggestion(BaseModel):
    """Live deal scraped from Canadian Tire for offer creation intelligence."""

    deal_id: str
    product_name: str
    category: str
    discount_pct: float = Field(..., ge=0, le=100)
    original_price: float = Field(..., ge=0)
    deal_price: float = Field(..., ge=0)
    source_url: str
    source: str = Field(..., pattern="^(weekly_deals|flyer|clearance)$")
    suggested_objective: str
    scraped_at: datetime = Field(default_factory=datetime.utcnow)
